internal/storage/metrics/postgres/postgres.go

```diff
@@ -0,0 +1,135 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"log"
+	"sync"
+
+	"github.com/JohnRobertFord/go-plant/internal/config"
+	"github.com/JohnRobertFord/go-plant/internal/storage/metrics"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+const (
+	createTableQuery = `CREATE TABLE IF NOT EXISTS metrics(
+	"id" int generated always as identity,
+	"name" varchar(255) UNIQUE,
+	"type" varchar(50),
+	"value" double precision,
+	"delta" bigint
+	);`
+	insertWithConflictQuery = `INSERT INTO metrics(name, type, value, delta) VALUES($1,$2,$3,$4) ON CONFLICT (name) DO UPDATE SET value=$5, delta=$6 ;`
+	getAllMetricsQuery      = `SELECT name, type, value, delta FROM metrics;`
+	getOneMetricQuery       = `SELECT name, type, value, delta FROM metrics WHERE name=$1 AND type=$2;`
+)
+
+type postgres struct {
+	db  *pgxpool.Pool
+	cfg *config.Config
+}
+
+var (
+	pgInstance *postgres
+	pgOnce     sync.Once
+)
+
+func (p *postgres) Close() {
+	p.db.Close()
+}
+
+func NewPostgresStorage(c *config.Config) *postgres {
+	ctx := context.Background()
+	pgOnce.Do(func() {
+		dbPool, err := pgxpool.New(ctx, c.DatabaseDsn)
+		if err != nil {
+			log.Printf("unable to create connection pool: %v", err)
+			return
+		}
+		pgInstance = &postgres{dbPool, c}
+	})
+	pgInstance.db.Exec(ctx, createTableQuery)
+
+	return pgInstance
+}
+
+func (p *postgres) SelectAll() []metrics.Element {
```

Comment:
```suggestion
func (p *postgres) SelectAll(ctx context.Context) ([]metrics.Element, error) {
```
Контекст следует брать из запроса, в обратную сторону обязательно возвращать ошибку. Наверху обрабатывать. И так для всех функций стораджа.